### PR TITLE
Integrate the work for hoisting address computation

### PR DESCRIPTION
**Pull request by qcolombet**
Add a pass to extract address computation from memref.load and nvgpu.ldmatrix.
Plumb the affine.apply decomposition through a new pass: decompose-affine-ops.
Rework the lowering pipeline to connect all the piece together:
1. extract-address-computation turns address computation into subviews
2. expand-strided-metadata turns subviews into affine.apply
3. licm hoists the code introduced by # 2 in the right scf.for loop
4. decompose-affine-ops breaks down the `affine.apply`s so that the resulting
   subexpressions can be hoisted in the right loops.
5. licm hoists the code introduced by # 4
6. lower-affine materializes the decomposed `affine.apply`s. We do that early
   to avoid the canonicalization to undo this work.

Phase 3-5 needs to run on `scf.for`, so the whole process has to run before
scf to cf.

Missing bits:
- Add support for memref.store, vector.transfer_xxx
- LLVM reassociate undoes some of the thing we improve here. Need to investigate and fix (tracked in https://github.com/llvm/llvm-project/issues/61458). 
- We need Nicolas' https://reviews.llvm.org/D145685 to be integrated to be able to land this.

Note: extract-address-computation could be moved to LLVM open source,
but we need to figure out where it could live since it has both a dependency on
memref and nvgpu. We probably want to come up with an interface like
`isAddressComputationExtractable` to push it upstream.

@nicolasvasilache for the decompose-affine part, I went with a pass in iree for now. We need to sync-up on how we could tackle what you mentioned in https://reviews.llvm.org/D145685#inline-1410000.
Also I pulled https://reviews.llvm.org/D145801 in IREE for now (the extract-address-computation pass). I think we'll want to go with an interface for the rewrite upstream but I wanted to have something working in this PR to unblock other investigations.

### CODE REVIEW

**Comment by qcolombet** [comment]
Filed https://github.com/llvm/llvm-project/issues/61458 for the issue with the reassociate pass in the LLVM backend.

File: compiler/src/iree/compiler/Codegen/Passes.td [hunk 1]
```diff
@@ -47,6 +47,28 @@ def DecomposeLinalgGeneric :
...
+}
+
+def ExtractAddressComputation: Pass<"extract-address-computation"> {
+  let summary = "TODO";
+  let description = [{
+     TODO
+  }];
+  let constructor = "mlir::iree_compiler::createExtractAddressComputationPass()";
+  let dependentDialects = [
+      "memref::MemRefDialect", "nvgpu::NVGPUDialect"
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Passes.td]: Could we decouple the dependence on `NVGPUDialect`. So a pass that can run on CPU without that dependency, and a pass with potentially extra patterns that are needed for `NVGPU`

Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Passes.td]: Sure thing

**Comment by qcolombet** [comment]
How can I add a test for `iree-compile` to make sure that the pass pipeline is hooked up properly?

Should I use `mlir-print-ir-after=xxx`, e.g., on the test case from https://github.com/openxla/iree/issues/12517 ?

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 1 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Passes.td]: I've split the patterns into two passes now.
The GPU pass is a super set of the "generic" pass for now.

Let me know if that works or if you prefer to have the generic and GPU pass being disjointed.

Also let me know if you want them to live in different files.

Finally, I didn't know if it was okay to leave the GPU one in `Common`, but saw that some GPU passes were already there (e.g., `GPUPipelining.cpp`).

**Comment by qcolombet** [comment]
I need help for the following two things:
- A test of `iree-compile` (what I mentioned in https://github.com/openxla/iree/pull/12643#issuecomment-1473899637)
- How to get the right sizes out of a `ldmatrix` (see https://github.com/openxla/iree/blob/9f5bdee0e5195fb0475245f46d507a6aa417a746/compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputationPass.cpp#L106)

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputationPass.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+static void populateExtractAddressComputationPatterns(
+    RewritePatternSet &patterns) {
+  patterns.add<
+      LoadLikeOpRewriter<memref::LoadOp,
+                         /*getSrcMemRef=*/getLoadOpSrcMemRef,
+                         /*rebuildOpFromAddressAndIndices=*/rebuildLoadOp>>(
+      patterns.getContext());
+}
+
+static void populateExtractAddressComputationGPUPatterns(
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputationPass.cpp]: Sorry for being annoying about this, but could we move these patterns into `LLVMGPU`. No other backends need this (or can use this AFAIK. I dont know SPIR-V use these as well. If that is the case, then it is fine here).

**Review by MaheshRavishankar — COMMENTED** [review]
This looks great.... You can add the benchmark labels on these just to see the impact of this.

**Comment by MaheshRavishankar** [comment]
> How can I add a test for `iree-compile` to make sure that the pass pipeline is hooked up properly?
> 
> Should I use `mlir-print-ir-after=xxx`, e.g., on the test case from #12517 ?

You can use `iree-compile` and `mlir-print-ir-after`... but you might be better of starting with an example in HAL. So you take the dispatch that HAL sees and do something like

```
iree-opt --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(iree-llvmcpu-lower-executable-target/iree-llvmgpu-lower-executable-target)))" %s | FileCheck %s
```

If that is too heavy weight you can just add a new pipeline that just has the new passes you added and check that...

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputationPass.cpp (hunk 2 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputationPass.cpp]: Sure thing!
I'll follow your direction, I don't yet understand the directory layout in IREE.

**Comment by qcolombet** [comment]
> > How can I add a test for `iree-compile` to make sure that the pass pipeline is hooked up properly?
> > Should I use `mlir-print-ir-after=xxx`, e.g., on the test case from #12517 ?
> 
> You can use `iree-compile` and `mlir-print-ir-after`... but you might be better of starting with an example in HAL. So you take the dispatch that HAL sees and do something like
> 
> ```
> iree-opt --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(iree-llvmcpu-lower-executable-target/iree-llvmgpu-lower-executable-target)))" %s | FileCheck %s
> ```
> 
> If that is too heavy weight you can just add a new pipeline that just has the new passes you added and check that...

That sounds perfect. Thanks for the suggestion.

**Comment by qcolombet** [comment]
> > > How can I add a test for `iree-compile` to make sure that the pass pipeline is hooked up properly?
> > > Should I use `mlir-print-ir-after=xxx`, e.g., on the test case from #12517 ?
> > 
> > 
> > You can use `iree-compile` and `mlir-print-ir-after`... but you might be better of starting with an example in HAL. So you take the dispatch that HAL sees and do something like
> > ```
> > iree-opt --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(iree-llvmcpu-lower-executable-target/iree-llvmgpu-lower-executable-target)))" %s | FileCheck %s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If that is too heavy weight you can just add a new pipeline that just has the new passes you added and check that...
> 
> That sounds perfect. Thanks for the suggestion.

I went with the pipeline involving `iree-codegen-linalg-to-nvvm-pipeline` instead of `iree-llvmgpu-lower-executable-target` because the original test case went through that (via `CUDATargetBackend::buildTranslationPassPipeline`).

File: compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+}  // namespace
+
+void DecomposeAffineOpsPass::runOnOperation() {
+  IRRewriter rewriter(&getContext());
+  this->getOperation()->walk([&](AffineApplyOp op) {
+    rewriter.setInsertionPoint(op);
+    reorderOperandsByHoistability(rewriter, op);
+    (void)decompose(rewriter, op);
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp:28]: nit: should we check the result and signal pass failure if it fails?
```
      if (failed(decompose(rewriter, op)) {
        return signalPassFailure();
      }
```

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    // If our load already has only zeros as indices there is nothing
+    // to do.
+    SmallVector<OpFoldResult> indices =
+        getAsOpFoldResult(loadLikeOp.getIndices());
+    if (std::all_of(indices.begin(), indices.end(),
+                    [](const OpFoldResult &opFold) {
+                      std::optional<int64_t> maybeConstant =
+                          getConstantIntValue(opFold);
+                      return maybeConstant && *maybeConstant == 0;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h]: nit: you could use `isConstantIntValue` helper:
https://github.com/llvm/llvm-project/blob/fa326c21919faea33b0470b7083dccbd3c1fc3f2/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h#L82

File: compiler/src/iree/compiler/Codegen/Common/test/decompose_affine_ops.mlir [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// of the loop: IV2 -> IV1 -> IV0
+// CHECK: #[[$SINGLE_VALUE_MAP:.*]] = affine_map<()[s0] -> (s0)>
+// CHECK: #[[$MUL_MAP:.*]] = affine_map<()[s0, s1] -> (s1 * s0)>
+// CHECK: #[[$ADD_MAP:.*]] = affine_map<()[s0, s1] -> (s0 + s1)>
+// CHECK-LABEL:   func.func @decomposeAffine(
+// CHECK-SAME:                            %[[VAL_0:.*]]: memref<?x?x?xf32, strided<[?, ?, ?], offset: ?>>) -> f32 {
+// CHECK:           %[[VAL_1:.*]] = arith.constant 0.000000e+00 : f32
+// CHECK:           %[[VAL_2:.*]] = arith.constant 0 : index
+// CHECK:           %[[VAL_3:.*]] = arith.constant 1 : index
+// CHECK:           %[[VAL_4:.*]] = arith.constant 2 : index
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/test/decompose_affine_ops.mlir]: use `CHECK-DAG`

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 6]
```diff
@@ -486,6 +506,10 @@ static void addLowerToLLVMGPUPasses(OpPassManager &pm, bool useROCM) {
 
   pm.addNestedPass<func::FuncOp>(createLLVMGPUVectorLoweringPass());
 
+  // THIS NEEDS TO RUN BEFORE SCF ->CF ON
+  addLowerAndOptimzeAddressComputation(pm);
+  // THIS NEEDS TO RUN BEFORE SCF ->CF OFF
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:513]: Could this run before as well? Like before software pipelining when doing matmul code?

File: compiler/src/iree/compiler/Codegen/Passes.td (hunk 1 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Passes.td]: Yes most of things are can be applied to both LLVMGPU and SPIRV are in `Common` right now. We discussed splitting it into a GPUCommon folder but we haven't got to it so far.

**Review by ThomasRaoux — COMMENTED** [review]
Sorry for the delayed review. This is awesome, thanks @qcolombet !
I added a few minor comments/questions.

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h (hunk 4 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h]: Ah nice, I missed that.
Thanks!

File: compiler/src/iree/compiler/Codegen/Common/test/decompose_affine_ops.mlir (hunk 5 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/test/decompose_affine_ops.mlir]: Good point!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 6 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:513]: That I don't know.
This lowering will turn affine into llvm stuff and also we need to have all the memrefs stuff (i.e., after bufferization).
So I don't know how much flexibility there is here and if the pipelining can work within these constraints.
You tell me :).

Where do you think we should move it?

File: compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp (hunk 3 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp:28]: I don't think we want to signal a failure when this "fails".
The failures happen when we encounter `affine.apply`s that cannot be rewritten (https://github.com/llvm/llvm-project/blob/fa326c21919faea33b0470b7083dccbd3c1fc3f2/mlir/include/mlir/Dialect/Affine/Transforms/Transforms.h#L37) and for these I think it is okay to just skip them.

**Review by nicolasvasilache — APPROVED** [review]
Generally LGTM and in line with what we discussed, great stuff!

As a followup, we could also pick some of the old LLVM canonicalization patterns I added a while back and port them to extract_strided_metadata to reduce the number of spurious accesses to structs: all these things should just canonicalize at a higher level now and be hoisted properly.

**Comment by qcolombet** [comment]
> This is awesome, thanks @qcolombet !

Thanks, but to be fair the heavy lifting was done by @nicolasvasilache for all the decompose affine and such!

> Generally LGTM and in line with what we discussed, great stuff!

Thanks, I'm just making what you did shine :).

> As a followup, we could also pick some of the old LLVM canonicalization patterns I added a while back and port them to extract_strided_metadata to reduce the number of spurious accesses to structs: all these things should just canonicalize at a higher level now and be hoisted properly.

Sounds good to me.
Do you have a pointer on that?
I think it would fit in a broader "let's rethink address lowering" effort.

**Comment by qcolombet** [comment]
> Regressed Latencies <...>

@MaheshRavishankar , @ThomasRaoux 

How do I reproduce that?
I'd like to make sure there's nothing fishy going on.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 6 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:513]: It will lower affine to arithmetic right? Not to LLVM dialect? lowering affine late in pipeline flow should be fine. The pipelining doesn't rely on affine to work. I would move it here to replace the licm pass:
https://github.com/openxla/iree/blob/ea0b28fe850a06123383cb1cf7671995362c06b1/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp#L329

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 7]
```diff
@@ -498,8 +522,6 @@ static void addLowerToLLVMGPUPasses(OpPassManager &pm, bool useROCM) {
   pm.addNestedPass<func::FuncOp>(createPolynomialApproximationPass());
 
   pm.addNestedPass<func::FuncOp>(arith::createArithExpandOpsPass());
-  pm.addNestedPass<func::FuncOp>(memref::createExpandOpsPass());
-  pm.addPass(memref::createExpandStridedMetadataPass());
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: I would leave those passes there as a fall back for now.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 6 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:513]: > It will lower affine to arithmetic right? Not to LLVM dialect? 

Yes, silly me.

**Comment by ThomasRaoux** [comment]
> > Regressed Latencies <...>
> 
> @MaheshRavishankar , @ThomasRaoux
> 
> How do I reproduce that? I'd like to make sure there's nothing fishy going on.

I used to get the artifact from buildkit to reproduce looks like things have moved recently though. Now you need to get the artifact as described here: https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/benchmark_suites.md#fetching-benchmark-artifacts-from-ci
Then you can compile and run iree-benchmark-module. That being said I have not done it through this new flow so there may be rough edges I don't know about.
I would suggest limiting your new transformation to the tensorcore matmul pipelines (there are 2) for now and mot likely this will prevent any unwanted behavior that could cause regressions.

**Comment by MaheshRavishankar** [comment]
> > > Regressed Latencies <...>
> > 
> > @MaheshRavishankar , @ThomasRaoux
> > 
> > How do I reproduce that? I'd like to make sure there's nothing fishy going on.
> 
> I used to get the artifact from buildkit to reproduce looks like things have moved recently though. Now you need to get the artifact as described here: https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/benchmark_suites.md#fetching-benchmark-artifacts-from-ci
> Then you can compile and run iree-benchmark-module. That being said I have not done it through this new flow so there may be rough edges I don't know about.
> I would suggest limiting your new transformation to the tensorcore matmul pipelines (there are 2) for now and mot likely this will prevent any unwanted behavior that could cause regressions.

These instructions might help. https://github.com/openxla/iree/blob/main/benchmarks/README.md

This should be applicable everywhere. If you have trouble I can help getting some artifacts. It's be really sad if we restrict things to just matmul pipelines. We have other workflows as well, and try to lift all boats when possible.

**Comment by ThomasRaoux** [comment]
> 
> This should be applicable everywhere. If you have trouble I can help getting some artifacts. It's be really sad if we restrict things to just matmul pipelines. We have other workflows as well, and try to lift all boats when possible.

it only does something useful when there is serial loop and we want hoist pointer arithmetic out. It is currently mainly the case on matmul kernels. Since it shouldn’t be applied at lowering time but earlier in the pipeline (for all the reasons discussed above) it makes sense to start by enabling matmul first. I’m not saying it shouldn’t be used in other places if needed. What other case did you expect this to help with current state?

**Comment by MaheshRavishankar** [comment]
> > 
> > This should be applicable everywhere. If you have trouble I can help getting some artifacts. It's be really sad if we restrict things to just matmul pipelines. We have other workflows as well, and try to lift all boats when possible.
> 
> it only does something useful when there is serial loop and we want hoist pointer arithmetic out. It is currently mainly the case on matmul kernels. Since it shouldn’t be applied at lowering time but earlier in the pipeline (for all the reasons discussed above) it makes sense to start by enabling matmul first. I’m not saying it shouldn’t be used in other places if needed. What other case did you expect this to help with current state? 
> 

Any reduction kernel on CPU, and maybe even conv kernels that have a serial reduction loop?
In any case @qcolombet let me know if you need any help.

**Comment by pzread** [comment]
> > Regressed Latencies <...>
> 
> @MaheshRavishankar , @ThomasRaoux
> 
> How do I reproduce that? I'd like to make sure there's nothing fishy going on.

https://github.com/openxla/iree/blob/6b8aa0f503b49336d659024df94e222e311023d0/docs/developers/developing_iree/benchmark_suites.md#find-compile-and-run-commands-to-reproduce-benchmarks should give you the latest instructions from fetching artifacts to get the commands to reproduce a x86/CUDA benchmark locally. Let me know if it doesn't work

**Comment by ThomasRaoux** [comment]
> > > This should be applicable everywhere. If you have trouble I can help getting some artifacts. It's be really sad if we restrict things to just matmul pipelines. We have other workflows as well, and try to lift all boats when possible.
> > 
> > 
> > it only does something useful when there is serial loop and we want hoist pointer arithmetic out. It is currently mainly the case on matmul kernels. Since it shouldn’t be applied at lowering time but earlier in the pipeline (for all the reasons discussed above) it makes sense to start by enabling matmul first. I’m not saying it shouldn’t be used in other places if needed. What other case did you expect this to help with current state?
> 
> Any reduction kernel on CPU, and maybe even conv kernels that have a serial reduction loop? In any case @qcolombet let me know if you need any help.

I was talking about GPU codegen here. The comment was about where to put it in the GPU codegen flow.

**Comment by qcolombet** [comment]
Thanks all for the repro steps.

Quick update here: out of the 14 dispatches in the regressing benchmark (clip text), only 4 are affected (number 10-13).
Looking at the diffs in the PTXs, there is nothing unexpected:
- We hoist more stuff
- Register allocation changed
- Slight various in instruction schedule

On the "hoist more stuff", this can affect the register pressure and maybe this is pushing us above an occupancy cliff.

The only surprising effect is in dispatch 11 where the final layout of the basic blocks is different because some conditions got flipped:
Before:
```asm
add.s64         %rd150, %rd150, 1;
add.s64         %rd149, %rd149, 308;
setp.eq.s64     %p5, %rd150, %rd8;
@%p5 bra        $L__BB0_2; 
// fallthrough
```

After:
```asm
add.s64         %rd157, %rd157, 1;
add.s64         %rd156, %rd156, 308;
setp.ne.s64     %p5, %rd157, %rd7;
@%p5 bra        $L__BB0_3;
bra.uni         $L__BB0_7; 
```

And in the end we end up with one more uniform branch and one more predicated branch.

I'm profiling with ncu to see what is causing the problem, but if my observations turned to be the problem it sounds to me that there is nothing to fix in this particular PR.

**Comment by ThomasRaoux** [comment]
> Thanks all for the repro steps.
> 
> Quick update here: out of the 14 dispatches in the regressing benchmark (clip text), only 4 are affected (number 10-13). Looking at the diffs in the PTXs, there is nothing unexpected:
> 
> * We hoist more stuff
> * Register allocation changed
> * Slight various in instruction schedule
> 
> On the "hoist more stuff", this can affect the register pressure and maybe this is pushing us above an occupancy cliff.
> 
> The only surprising effect is in dispatch 11 where the final layout of the basic blocks is different because some conditions got flipped: Before:
> 
> ```assembly
> add.s64         %rd150, %rd150, 1;
> add.s64         %rd149, %rd149, 308;
> setp.eq.s64     %p5, %rd150, %rd8;
> @%p5 bra        $L__BB0_2; 
> // fallthrough
> ```
> 
> After:
> 
> ```assembly
> add.s64         %rd157, %rd157, 1;
> add.s64         %rd156, %rd156, 308;
> setp.ne.s64     %p5, %rd157, %rd7;
> @%p5 bra        $L__BB0_3;
> bra.uni         $L__BB0_7; 
> ```
> 
> And in the end we end up with one more uniform branch and one more predicated branch.
> 
> I'm profiling with ncu to see what is causing the problem, but if my observations turned to be the problem it sounds to me that there is nothing to fix in this particular PR.

What kind of op is this coming from? As mentioned I would suggest limiting to matmul pipeline for now to avoid such problems. In addition the regressing benchmark is new and not something we have profiled and made sure the code generated is reasonable so we might be fine with the regression if there is no other solutions.
However I assume this is will go away once we only apply this to only selected pipelines it will go away

File: compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp:28]: Its OK for doing that now, but that might mean the `decompose` method needs to return state better. If it is not a real failure that needs to be caught by the compiler, it shouldnt return failure. Would be good to fix (but I dont care enough to make it required)

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+struct ExtractAddressComputationPass
+    : public ExtractAddressComputationBase<ExtractAddressComputationPass> {
+  void runOnOperation() override;
+};
+}  // namespace
+
+void ExtractAddressComputationPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
+  populateExtractAddressComputationPatterns(patterns);
+  (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.cpp]: This would be good to catch and signal failure. This typically returns failure when pattern application fails to converge, which should be an error and better to catch that.

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+/// %new_base = subview %base[%off0,.., %offN][1,..,1][1,..,1]
+/// %ld = memref.load %new_base[0,..,0] :
+///    memref<1x..x1xTy, strided<[1,..,1], offset: ?>>
+///
+/// `getSrcMemRef` returns the source memref for the given load-like operation.
+///
+/// Using the given rewriter, `rebuildOpFromAddressAndIndices` creates a new
+/// LoadLikeOp that reads from srcMemRef[indices].
+/// The returned operation will be used to replace loadOp.
+template <typename LoadLikeOp, Value (*getSrcMemRef)(LoadLikeOp),
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h]: Would it be possible to not make this templated and move the implementation out of a header file? Fine if this can be done later as well. Seems like this is calling for a `LoadOpInterface`. So not a blocking requirement.

**Review by MaheshRavishankar — APPROVED** [review]
This only changing the GPU codegen. Did you get a chance to just add this to the CPU codegen pipelines? We probably have similar issues there. If you are planning to do that as a follow up thats fine, but would be great to file an issue for enabling these on other backends as well.
I just have one comment on handling of return of `applyPatternsAndFoldGreedily`, but stamping since I dont have any other comments.,

**Comment by qcolombet** [comment]
> What kind of op is this coming from?

I haven't checked, I would expect a regular `scf.if` and the backend choose different heuristics for the basic block layout.
I haven't tried to track that down because the IR is massive and printing it takes forever.

Agree that we could limit it to the matmul pipeline for now, but I'd like to understand what is going on to at least track what we need to fix that (or not) at some point. And to Mahesh's point this should be generally beneficial.

Anyway, now that I have the profiles, I can confirm that dispatch 11 is the problem. It went from 26 µs to 98 µs!
Looking at the profile, what stands out is the new codegen does a lot more LDG and STG.
LDG went from 221K to 420K
STG went from 12K to 183K!!!

The LDG could have been expected given the hoisting, though I would have expected the backend to rematerialize them if the estimated frequencies are so much different. The STG are just puzzling to me, in particular we have as many of them in the PTXs and although it is not immediately obvious I think they are in the same spot in the CFG. I'll double check that part.

Looking at the final SASS I wonder if the backend is not taking some weird path in its heuristics. The input PTXs are both 256 LoC and differ only from the hoisting and basic block layout, but the final SASS are 600 instructions without this PR and 800 with. The loop body in particular is much bigger with a lot more STG.

Anyhow, I'll continue tomorrow.

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.cpp (hunk 8 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.cpp]: Good point.
I was only thinking in terms of optimization: this optimization is optional so if it fails we don't care, but actually we may not be converging like you said (albeit it cannot happen right now :)).

I'll fix that.

File: compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h (hunk 9 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h]: Yes, the plan is to add an interface and get rid of this.
The interface itself will like spark some conversation in LLVM upstream and I didn't want to wait for that to happen.

**Comment by qcolombet** [comment]
> Did you get a chance to just add this to the CPU codegen pipelines?

No I did not but I expect this would help here as well.
My plan was to land this in the GPU to unlock the matmul improvemens, properly refactor all that in upstream LLVM, then get back that in IREE for all backends.

**Comment by qcolombet** [comment]
> I assume this is will go away once we only apply this to only selected pipelines it will go away

Concretely @ThomasRaoux this means moving the call to `addLowerAndOptimzeAddressComputation` in both `addGPUMatmulTensorCorePassPipeline` and `addGPUMatmulTensorCoreMmaSyncPassPipeline`, right?

**Comment by ThomasRaoux** [comment]
> > I assume this is will go away once we only apply this to only selected pipelines it will go away
> 
> Concretely @ThomasRaoux this means moving the call to `addLowerAndOptimzeAddressComputation` in both `addGPUMatmulTensorCorePassPipeline` and `addGPUMatmulTensorCoreMmaSyncPassPipeline`, right?

yes

File: compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp (hunk 3 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/Common/DecomposeAffineOpsPass.cpp:28]: @nicolasvasilache what do you think?

I’m thinking that maybe we want to return a simple bool that would report if anything was changed.

**Comment by qcolombet** [comment]
Went to the bottom of the regression.
**The TL;DR is we were missing to apply the rewrite pattern to stores and that confused later optimizations**

Now for the details: what happened is the stores (and some loads) were actually not in the same spot in the CFG.
Without this PR:
```asm
        @%p12 bra       ...;
        st.global.f32   [%rd57], %f10;
        add.s64         %rd170, %rd170, 1;
```
With
```asm
        st.global.f32   [%rd57], %f10;
        @%p12 bra       ...;
        add.s64         %rd170, %rd170, 1;
```

Although the basic block is not labeled, the previous code had the store in the exit block of the loop whereas the new code had the store in the loop.

The optimization doing this code motion is LICM in the LLVM backend. The pattern that it recognizes is:
```
loop:
  val = ld addr
  new_val = ... // do some stuff with val
  store new_val, addr
```
And it turns that into:
```
val = ld addr
loop:
  phi_val = phi(val, new_val)
  new_val ... // do some stuff with phi_val
exit:
store new_val, addr
```

Now this PR was messing with this optimization because we were rewriting the address of the load, but not of the store. As a result, while both addresses are still the same, given the sub-expressions were ordered differently they were optimized differently and it was difficult to see that they were still the same.

To fix that, I added the rewrite of the store operation as well so the addresses stay in sync.

**Comment by nicolasvasilache** [comment]
Great, thanks for digging and adding this!

**Comment by qcolombet** [comment]
@ThomasRaoux should we go with the current PR?
I know you tried to move the sub-pipeline earlier and it was causing some perf issues. Therefore that sounds like something we should revisit later, in particular it could fit in how do we enable this for all backends.

**Comment by ThomasRaoux** [comment]
> @ThomasRaoux should we go with the current PR?
> I know you tried to move the sub-pipeline earlier and it was causing some perf issues. Therefore that sounds like something we should revisit later, in particular it could fit in how do we enable this for all backends.

Sounds good, yes let’s go with this for now.

**Review by ThomasRaoux — APPROVED** [review]
